core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala

```diff
@@ -201,6 +332,25 @@ private[spark] class ExecutorMonitor(
     }
   }
 
+  override def onOtherEvent(event: SparkListenerEvent): Unit = event match {
+    case ShuffleCleanedEvent(id) => cleanupShuffle(id)
+    case _ =>
+  }
+
+  override def rddCleaned(rddId: Int): Unit = { }
+
+  override def shuffleCleaned(shuffleId: Int): Unit = {
+    // Because this is called in a completely separate thread, we post a custom event to the
+    // listener bus so that the internal state is safely updated.
+    listenerBus.post(ShuffleCleanedEvent(shuffleId))
```

Comment:
it is weird that this will be visible to all user listeners as well ...

Reply:
Well, all listeners already have to deal with events they don't understand; that's the contract for `onOtherEvent`, since we're allowed to add new events (public or not).

The only odd thing here is that this event is not public, so you can't really handle it outside of Spark (without reflection), but I don't really see an easy way to solve this issue differently (outside of adding locking which I don't want to).

Reply:
yeah, that's the weird part, you get this event with a class that is not public.

The only things I can think of are (1) to create a new event loop here, which just forwards the listener events and also this new event or (2) create a "InternalListenerEvent" which is automatically filtered from all non-internal listeners.  But both of those seem like overkill.